### DISCUSSION TITLE
Tips for using this in a real-time simulation environment

### DISCUSSION THREAD

**Discussion by ziriax** [jscad/OpenJSCAD.org#1308] [General]
I'm planning to use this library in my never-ending never-completed pet project https://vikid.net

The main idea is that just like with image processing authoring tools, you can visualize every intermediate output, on the fly, which greatly helps with debugging and reasoning in general (at least once your brain gets used to it, initial users often find it too much information displayed at once)

So ViKiD is all about real-time rendering of many graphics in many canvasses... 

I think OpenJSCAD is mostly targetted towards off line generation of 3D CAD models, right?

But with clever caching, would it be possible to use it in a real-time simulation, as long as the 3D model generation parameters don't change (e.g. only doing transformation, not deformation or construction)?

For example, I found this https://www.npmjs.com/package/@jscad/vtree experimental package, which seems to go into that direction?

I was first planning to do everything myself using webgpu, doing all the CSG on the GPU, but as an NVIDIA software engineer, I just don't have enough free time. But some users really would like to see 3D in ViKiD, so I'm looking for possible solutions that fit into the functional dataflow mindset, and OpenJSCAD surely fits this paradigm perfectly (at least the "functional" part)

I first started to integrate [thaichi.js](https://taichi-js.com/playground/fractal-on-cloth), but that is also long term.

**Comment by z3dev** [comment]
@ziriax Very cool never-completed pet project... but it looks pretty complete to me. ;)

> I think OpenJSCAD is mostly targetted towards off line generation of 3D CAD models, right?

Actually, just the opposite. JSCAD is written in 100% Javascript, and all packages are created to run in browsers. You can find everything on-line at www.openjscad.xyz

I took a look at the the VIKID website. I think JSCAD would integrate very nicely. There are primitives as well as operations, and of course math as well. Everything is available via functions as well. There's no 'object' hierarchy to learn.

V3 is going every further. All the packages are moving to ES packaging, which should make the packages very easy to embed in websites. And @hrgdavor @platypii are putting some finishing touches on the new UI parts for the new website.

Caching... yeah. There are all kinds of caching.

VTREE is meant to speed up the calculations of pieces, in JSCAD terms. But VTREE would need some major rework now.

Another area of caching would be in the graphics. It's fairly easy to convert from JSCAD to webGL buffers, especially when using Three.js or other graphics libraries. And once converted, those graphic entities can be moved as well.

So, again. I would say that JSCAD is a good starting place to integrate 3D shapes into VIKID.net

**Reply by ziriax** [reply]
My bad, I said "off-line" but I meant "not real-time"

Thank you, I will give it a shot!
